### ISSUE TITLE
Include Account Transfers

### ISSUE THREAD

**Issue by divegeek** [divegeek/moneydance-detailedbudget#2]
Note:  Imported from http://code.google.com/p/moneydance-detailedbudget/issues/detail?id=2

What steps will reproduce the problem?
1. Generating a detailed budget when part of the budget includes account
transfers (e.g. an Escrow account as part of a mortgage payment or the
principal portion of a mortgage payment that is a "transfer" to a mortgage
loan account in MD) 

What is the expected output? What do you see instead?
As mentioned above, I personally have a couple of items that show up in my
budget that aren't set up as expense categories but rather are account
transfers. Two specific examples are the Escrow Transfer and Principal
portion of my mortgage payments. The standard budget feature of MD includes
these in the report because I have included the transfers as part of my
budget. It would be nice to have them show up in the detailed budget report
as well.

What version of the product are you using? On what operating system?
Moneydance 2010r2 (build 735) w/latest available Detailed Budget extension
(v 0.5 I believe)

Please provide any additional information below.

**Comment by divegeek** [comment]
I'll look into this.  I have some budget items that are account transfers as well, and I'd also like this behavior.

It's not clear to me how this should work, exactly, though.  Take your example of an escrow account:

There's a monthly transfer TO the escrow account that is part of the monthly mortgage payment split, but there are also a couple of transfers FROM the escrow account annually, when the insurance and property tax payments are made.  Should the extension use both the TO and FROM transfers, summing them all?  Over the course of a year, then, your escrow account budget will sum to zero (ideally).  That doesn't seem like what you want.

On the other hand, if the extension looks only at the TO transfers, that will (a) be inconsistent with the way it works for categories and (b) not work right for another use case I have, which is for an investment account.  I have a monthly budgeted transfer to put money into my investment account.  But sometimes I pull money out of that account, and I want those FROM transfers to be included as well, because they represent negative progress on my savings goal.  If I put $5 in my investment account, but pull $1 out, then I want the report to show that I have put $4 in, falling short of my goal of $5.

Thoughts?

**Comment by therevruss** [comment]
Thanks for your looking into this and your questions. It sounds like we're on the same page here. In my example you're correct, I would only want to include the TO transfers, not necessarily the FROM. However, as you've suggested, this might be different for different accounts. With an Escrow account, it would be rare that I would ever transfer FROM escrow TO another account like Checking. With an investment account, like you mentioned, where TO and FROM transfers would happen regularly, I would expect MD to account for both. With an escrow account, I would have the property taxes and insurance as you stated. But now I'm realizing it's not as simple as I thought. For me, while I track my property tax and insurance payments for tax purposes, I don't include them in my monthly/yearly budget because they're wrapped into my escrow portion of my mortgage payment (which is why you said ideally the Escrow account balance should zero out). But,I would not want my escrow BUDGET to show zero dollars for the year. I'm sorry if I'm making this confusing - again, I think we're on the same page but let me just use some example numbers.

Mortgage Payment: $1000 ($500 Interest, $250 escrow, $250 principal) - I have both a mortgage account and escrow account in Moneydance and it sets up both the escrow and principal portions as transfers FROM checking TO those respective accounts.

For my annual budget: $6000 Interest (as a category), $3000 escrow (xfer FROM checking TO escrow), $3000 principal (xfer FROM checking TO mortgage account).

As the extension stands now, with my budget set up like this - it only seems to track the Interest portion of my payment as that is the only "category" and the other two are "accounts".

I'm guessing you already knew all that. Again, I just wanted to make sure I was being clear. It sounds like the real issue is the way MD handles categories - it actually treats them as separate accounts. In other words, even if the property tax payments and insurances payments in my escrow account are set up as "categories", MD still views them as a "transfer" - am I understanding this correctly? Furthermore, I would find my numbers would be off even with the MD basic budget function if I were to ever do a transfer from my escrow account back to say my checking.

**Comment by divegeek** [comment]
Yes, we're on the same page, and it sounds like the solution that I was contemplating is exactly the one that Moneydance is using for its budget display -- If you have a budget item that's an account, rather than a category, then transfers between that account and other accounts are counted, both TO and FROM, but expenditures TO an expense category are not.

Besides the fact that it's a logical solution, the detailed budget calculator really needs to mirror the functionality of the built-in budget displays/reports.  It would be very confusing if they worked differently.

I'm going to do some experiments to see if I can work out exactly how MD's budget reports work with account budget items.

BTW, my approach to budgeting for my mortgage is simpler and rougher than yours.  I don't actually budget each of the categories (interest, principal, escrow), I just budget the whole payment amount, and then I don't bother tracking actuals against budget, because there's really no way I could be either under or over budget, and because I really can't budget interest and principal amounts very effectively, since they change every month.  I could figure out a monthly amount for each that would work out to the correct figures for each year, I suppose, but I don't bother.

Still, it'd be nice if this report did work for this use case.

**Comment by divegeek** [comment]
A little experimentation shows that MD's budget report doesn't work the way we're thinking, and I think I've found some bugs.

It looks like with MD if the budget category is an account, only transfers into the account are counted towards the budget.  Any transfer out of the account is not counted, regardless if it's to another account or a payment to an expense category.  This seems to be the case regardless of the type of account (Bank/Asset/Liability/Loan/etc.).

I'm not sure if that's really ideal, but I think it makes sense to mirror it while I talk to Sean Reilly about whether or not it's what MD should be doing.

I've also noticed that with one of my investment accounts, many of my transfers into the account are not counted in the budget.  But it's working fine on other accounts.  That's where I think I've hit a bug.

**Comment by therevruss** [comment]
If it helps, the reason I do my mortgage the way I do is for tax purposes. I need to keep track of my principal, interest, and escrow payments separately so at the end of the year I know how much when to each category. So it's not so much that it's important for budgeting purposes - i.e., as you mentioned, they aren't going to change drastically over the year, but I can't just do 1 budget category because I need to know total amounts for each. In addition, I need an separate escrow account so I can use MD to track my property tax and insurance payments. That way at the end of the year I can just run a report and quickly see how much went to mortgage interest, insurance, property tax, and principal payment. So that's why I want the budget feature to only track the TO transfers for my mortgage account. 

From your comments, it sounds like it's either a bit more complicated than I assumed because your extension needs to mirror MD's basic budget report and/or there's a bug in the way MD reports budget figures. 

If there's any additional questions I can answer or insight I could provide that would be helpful to you, please let me know and I'll be happy to assist as I can.

**Comment by divegeek** [comment]
So, it's nearly a year later and I haven't made any progress on this issue.  In the interim I got a new job (a better one, but one that consumes more brain cycles), sold a house and bought another, relocated my family to a new state and much more, so... 

Anyway, I have a little time now and have been looking at some detailed budget stuff.  I notice that one of the things that has changed since Rolf wrote this extension that Moneydance now provides access to its own budget engine through the APIs.  I think I'm going to rewrite the extension using the new APIs.  On the one hand, this will ensure that the calculations match MD's budget bars.  On the other hand, it means that detailed budget will not be able to diverge from MD's method.

BTW, I also track my interest and escrow on my mortgage; each monthly payment is a split that allocates the money to the correct categories/accounts, and at the end of the year I can easily compare what I recorded that I paid for interest, etc. against what the bank reported.

I just don't bother breaking this out in the budget.  This means that if I select "show all" in detailed budget I see that I'm perpetually underspending on principal (since my whole payment is budgeted to that) and overspending on interest (since I'm not budgeting anything for that).  But budgeting the whole amount to one category is simple and accomplishes my main budgeting goal, which is to make it possible for me to verify that my budget "works", i.e. that income meets or exceeds expenses.  And the categorization of actuals means I can still get the tax figures I need.
